src/runtime-rs/crates/runtimes/virt_container/src/sandbox.rs

```diff
@@ -0,0 +1,228 @@
+// Copyright (c) 2019-2022 Alibaba Cloud
+// Copyright (c) 2019-2022 Ant Group
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+use std::sync::Arc;
+
+use agent::{self, Agent};
+use anyhow::{anyhow, Context, Result};
+use async_trait::async_trait;
+use common::{
+    message::{Action, Message},
+    Sandbox,
+};
+use containerd_shim_protos::events::task::TaskOOM;
+use hypervisor::Hypervisor;
+use kata_types::config::TomlConfig;
+use resource::{ResourceConfig, ResourceManager};
+use tokio::sync::{mpsc::Sender, Mutex, RwLock};
+
+use crate::health_check::HealthCheck;
+
+#[derive(Clone, Copy, PartialEq, Debug)]
+pub enum SandboxState {
+    Init,
+    Running,
+    Stopped,
+}
+
+struct SandboxInner {
+    state: SandboxState,
+}
+
+impl SandboxInner {
+    pub fn new() -> Self {
+        Self {
+            state: SandboxState::Init,
+        }
+    }
+}
+
+unsafe impl Send for VirtSandbox {}
+unsafe impl Sync for VirtSandbox {}
+#[derive(Clone)]
+pub struct VirtSandbox {
+    sid: String,
+    msg_sender: Arc<Mutex<Sender<Message>>>,
+    inner: Arc<RwLock<SandboxInner>>,
+    resource_manager: Arc<ResourceManager>,
+    agent: Arc<dyn Agent>,
+    hypervisor: Arc<dyn Hypervisor>,
+    monitor: Arc<HealthCheck>,
+}
+
+impl VirtSandbox {
+    pub async fn new(
+        sid: &str,
+        msg_sender: Sender<Message>,
+        agent: Arc<dyn Agent>,
+        hypervisor: Arc<dyn Hypervisor>,
+        resource_manager: Arc<ResourceManager>,
+    ) -> Result<Self> {
+        Ok(Self {
+            sid: sid.to_string(),
+            msg_sender: Arc::new(Mutex::new(msg_sender)),
+            inner: Arc::new(RwLock::new(SandboxInner::new())),
+            agent,
+            hypervisor,
+            resource_manager,
+            monitor: Arc::new(HealthCheck::new(true, true)),
+        })
+    }
+
+    async fn prepare_for_start_sandbox(
+        &self,
+        netns: Option<String>,
+        _config: &TomlConfig,
+    ) -> Result<Vec<ResourceConfig>> {
+        let mut resource_configs = vec![];
+
+        if let Some(_netns_path) = netns {
+            // TODO: support network
+        }
+
+        let hypervisor_config = self.hypervisor.hypervisor_config().await;
+        let virtio_fs_config = ResourceConfig::ShareFs(hypervisor_config.shared_fs);
+        resource_configs.push(virtio_fs_config);
+
+        Ok(resource_configs)
+    }
+}
+
+#[async_trait]
+impl Sandbox for VirtSandbox {
+    async fn start(&self, netns: Option<String>, config: &TomlConfig) -> Result<()> {
+        let id = &self.sid;
+
+        // if sandbox running, return
+        // if sandbox not running try to start sandbox
+        let mut inner = self.inner.write().await;
+        if inner.state != SandboxState::Init {
```

Comment:
This check conflicts with the comments above.

Reply:
Thanks for your review, has been updated.